### PR TITLE
Kuadrant performance test framework setup

### PR DESCRIPTION
**Pull request by dhlavac**
PR contains a basic setup of the Hyperfoil framework to test Kuadrant performance.  Hyperfoil allows to track and visualize response codes, latency distribution, and much more. See example reports below 
image
image
image

### CODE REVIEW

File: testsuite/oidc/rhsso/__init__.py [hunk 1]
```diff
@@ -85,3 +85,12 @@ def refresh_token(self, refresh_token):
     def get_token(self, username=None, password=None) -> Token:
         data = self.oidc_client.token(username or self.test_username, password or self.test_password)
         return Token(data["access_token"], self.refresh_token, data["refresh_token"])
+
+    def token_body_creation(self) -> str:
```
Comment (pehala) [inline testsuite/oidc/rhsso/__init__.py:89]: `token_params`?

I would also update the docstring to be more verbose

File: testsuite/perf_utils.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+def prepare_url(url: ParseResult) -> ParseResult:
+    """ Adds port number to url if it is not set"""
+    if not url.port:
+        url_port = 80 if url.scheme == 'http' else 443
+        url = url._replace(netloc=url.hostname + f':{url_port}')
+    return url
+
+
+class HyperfoilUtils:
```
Comment (pehala) [inline testsuite/perf_utils.py:33]: You could implement `LifecycleObject` to make it safer. (currently if something in init fails it wont be cleaned up)

File: testsuite/tests/kuadrant/authorino/performance/test_perf_basic.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+@pytest.fixture(scope='module')
+def number_of_agents():
+    """Number of spawned HyperFoil agents"""
+    return AGENTS
+
+
+@pytest.fixture(scope='module')
+def template(root_path):
+    """Path to template"""
+    return os.path.join(root_path, 'templates/template_perf_basic_query_rhsso.hf.yaml')
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/performance/test_perf_basic.py:29]: How about using `https://docs.python.org/3.11/library/importlib.resources.html`?

File: testsuite/tests/kuadrant/authorino/performance/conftest.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+
+@pytest.fixture(scope='session')
+def root_path():
+    """Root path for performance tests"""
+    return Path(os.path.realpath(__file__)).parent
+
+
+@pytest.fixture(scope="module")
+def rhsso_authorization(authorization, rhsso):
```
Comment (pehala) [inline testsuite/tests/kuadrant/authorino/performance/conftest.py:36]: Should be set up by default already, if you want to force RHSSO, only thing that's needed is to override `oidc_provider` fixture

File: config/settings.yaml [hunk 5]
```diff
@@ -10,4 +10,6 @@ default:
       username: "testUser"
       password: "testPassword"
   authorino:
-    deploy: true
\ No newline at end of file
+    deploy: true
+  hyperfoil:
+    url: "hyperfoil url"
```
Comment (pehala) [inline config/settings.yaml:15]: Probably shouldn't be defined here, but instead in `settings.local.yaml.tpl`

File: testsuite/perf_utils.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        """
+        Adds csv file with data for access token creation. Each row consists of following columns:
+        [authority url, rhsso url, rhsso path, body for token creation]
+        :param rhsso: rhsso service fixture
+        :param client_url: url of desired endpoint to be tested
+        :param filename: name of csv file
+        """
+        rows = []
+        token_url_obj = urlparse(rhsso.well_known['token_endpoint'])
+        token_port = 80 if token_url_obj.scheme == 'http' else 443
```
Comment (pehala) [inline testsuite/perf_utils.py:99]: Why are you not using `prepare_url`?

Comment (dhlavac) [reply testsuite/perf_utils.py:99]: good point, it was written before `prepare_url`

File: testsuite/tests/kuadrant/authorino/performance/test_perf_basic.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    """Init of hyperfoil utils"""
+    utils = HyperfoilUtils(hyperfoil_client, template)
+    request.addfinalizer(utils.finalizer)
+    return utils
+
+
+@pytest.fixture(scope='module')
+def setup_benchmark_rhsso(hyperfoil_utils, client, rhsso, number_of_agents):
+    """Setup of benchmark. It will add necessary host connections, csv data and files."""
+    url_pool = [{'url': rhsso.server_url, 'connections': 100}, {'url': str(client.base_url), 'connections': 20}]
```
[the remark this answers is not in the document]
Comment (dhlavac) [reply testsuite/tests/kuadrant/authorino/performance/test_perf_basic.py:43]: Added comment that mentions  numbers in connections

File: testsuite/tests/kuadrant/authorino/performance/conftest.py (hunk 4 above)
Comment (dhlavac) [reply testsuite/tests/kuadrant/authorino/performance/conftest.py]: Removed

File: Makefile [hunk 8]
```diff
@@ -41,6 +41,10 @@ test: ## Run test
 test pytest tests: pipenv
 	$(PYTEST) -n4 -m 'not flaky' --dist loadfile $(flags) testsuite
 
+# Run performance tests
+performance: pipenv
```
Comment (pehala) [inline Makefile:45]: Sadly, this still means `make test` will run performance tests, which we do not want

Comment (dhlavac) [reply Makefile:45]: Yeah sure 🤦 sorry

**Review by pehala — APPROVED** [review]

**Review by pehala — APPROVED** [review]

**Review by JaurbanRH — APPROVED** [review]

**Pull request merged**